### DISCUSSION TITLE
Freesql读写分离，读优先使用主库，主库不可用使用从库

### DISCUSSION THREAD

**Discussion by skyxioo** [dotnetcore/FreeSql#1382] [Q&A]
因为主从同步存在延时，很多场景写完之后需要立即查询结果，如果查从库，主从同步出现延时会导致数据不一致。
也许你可以说直接用：_freeSql.Select<T>().Master()。这里存在一个问题，主库不可用直接报异常，而不会自动切换从库。
在实际使用过程中也发现，在同一个事务中的查询，也不是查主库。
所以，我的需求是，freesql是否有这样的api或者配置，每次查询优先使用主库，主库不可用时用从库查。目前我的实现方式如下（感觉很别扭）：

public ISelect<T> SelectFsql<T>() where T : BaseEntity
 {
            // 因为Freesql是被动触发IsAvailable
            // 即使IsAvailable=true也非真的true，需要查一次
            if (_freeSql.Ado.MasterPool.IsAvailable)
                _freeSql.Ado.ExecuteConnectTest(1);

            // 优先使用主库，主库不可用使用从库
            return _freeSql.Ado.MasterPool.IsAvailable ? _freeSql.Select<T>().Master() : _freeSql.Select<T>();
 }

**Comment by 2881099** [comment]
你有几个从库，试试这样：

```c#
new FreeSqlBuilder()
    ....
    .UseConnectionString(DataType.MySql, master_connectionstring)
    .UseSlave(master_connectionstring, slave1_connectionstring) //把 master 也当成一个 slave 
    .UseSlaveWeight(100, 0)
    .Build();
```

**Reply by skyxioo** [reply]
这样我也试过，一样会读从库。我就一个主一个从

**Reply by skyxioo** [reply]
![图片](https://user-images.githubusercontent.com/29593571/209253126-aa189c47-3a16-4372-9967-4f23183791a1.png)
如图，结果是1的是主库，结果是0的是从库，从库出现的比例还很高，感觉UseSlaveWeight未生效一般
代码如下：
Fsql = new FreeSql.FreeSqlBuilder()
                .UseConnectionString(FreeSql.DataType.MySql, Configuration["ConnectionStrings:Master"])
                .UseSlave(Configuration["ConnectionStrings:Master"], Configuration["ConnectionStrings:Slave"])
                .UseSlaveWeight(100, 0)
                .Build();

**Reply by skyxioo** [reply]
_ = Task.Run(async () =>
            {
                while (true)
                {
                    var list = await _freeSql.Select<NorthGrant>()
                        .ToListAsync();
                    Console.WriteLine($"查询都记录数：{list.Count}");
                    Task.Delay(1000).Wait();
                }
            });

**Reply by 2881099** [reply]
哦，我知道了你测的是异步，异步代码确实不行。遗留问题，你试试同步

**Reply by 2881099** [reply]
异步该问题解决了，v3.2.686-preview20221224
